### PR TITLE
minimum distance constrain to fit_dipole

### PR DESCRIPTION
**Pull request by jmontoyam**
Added minimum distance constrain to fit_dipole: the distance from each fitted dipole to the inner skull should be greater or equal to some distance define  by the user.

### CODE REVIEW

**Comment by agramfort** [comment]
can you add a test?

File: mne/dipole.py [hunk 1]
```diff
@@ -338,10 +340,10 @@ def _fit_dipoles(data, times, rrs, guess_fwd_svd, fwd_data, whitener,
     return pos, amp, ori, gof, residual
 
 
-def _fit_dipole(B_orig, t, rrs, guess_fwd_svd, fwd_data, whitener, proj_op,
+def _fit_dipole(min_dist_to_inner_skull, B_orig, t, rrs,
+                guess_fwd_svd, fwd_data, whitener, proj_op,
                 fmin_cobyla):
     """Fit a single bit of data"""
-    logger.info('---- Fitting : %7.1f ms' % (1000 * t,))
```
Comment (Eric89GXL) [inline mne/dipole.py]: My poor logging statement!

File: mne/dipole.py [hunk 2]
```diff
@@ -354,7 +356,19 @@ def constraint(rd):
...
             else:
-                return 1.
+
+                # Once we know the dipole is below the inner skull,
+                # let's check if its distance to the inner skull is at least
+                # min_dist_to_inner_skull. If it is not, it is penalized
+                # proportional to its distance.
+
+                dist = _compute_nearest(surf['rr'], rd[np.newaxis, :],
+                                        return_dists=True)[1][0]
```
Comment (Eric89GXL) [inline mne/dipole.py]: Maybe computation should only be done if `dist is not None`, and `dist` should default to `None` for consistency with old behavior... I'm not sure. Depends on if this is considered a bug or not.

File: mne/dipole.py [hunk 3]
```diff
@@ -386,11 +400,18 @@ def constraint(rd):
...
+    dist = _compute_nearest(surf['rr'], rd_final[np.newaxis, :],
+                            return_dists=True)[1][0]
+    logger.info('---- Fitted : %7.1f ms, distance to inner skull : %2.4f mm' %
+                (1000 * t, dist * 1000))
     return rd_final, amp, ori, gof, residual
 
 
 @verbose
-def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
+def fit_dipole(evoked, cov, bem, trans=None, min_dist=5,
```
Comment (Eric89GXL) [inline mne/dipole.py]: ...but a min_dist of 1 should probably be sufficient, no?

File: mne/dipole.py [hunk 4]
```diff
@@ -404,6 +425,8 @@ def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
     trans : str | None
         The head<->MRI transform filename. Must be provided unless BEM
         is a sphere model.
+    min_dist : int, float
```
Comment (Eric89GXL) [inline mne/dipole.py]: just float

File: mne/dipole.py [hunk 5]
```diff
@@ -404,6 +425,8 @@ def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
...
+        Minimum distance (in milimeters) from the dipole to the inner skull.
```
Comment (Eric89GXL) [inline mne/dipole.py]: Maybe mention having it `> 0` is necessary to ensure numerical stability

File: mne/dipole.py [hunk 6]
```diff
@@ -429,6 +452,9 @@ def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
     times = evoked.times.copy()
     comment = evoked.comment
 
+    # Convert the min_dist to meters
+    min_dist_to_inner_skull = float(min_dist) / 1000
```
Comment (Eric89GXL) [inline mne/dipole.py]: this can be accomplished more simply as `min_dist / 1000.`

File: mne/dipole.py (hunk 3 above)
Comment (dgwakeman) [inline mne/dipole.py]: Not with any of our standard BEMs at least 5mm for our standard BEM. Ideally someone (not volunteering) would write a simple feature that calculates the average distance between neighboring vertices in the BEM and uses that to suggest the correct default. using the suggestions from Alex's OPENMEEG paper which I believe said for our type of BEM the BEM intervertex distance should be roughly equivalent to the mindist

File: mne/dipole.py (hunk 2 above)
Comment (agramfort) [inline mne/dipole.py]: +1

File: mne/dipole.py (hunk 3 above)
Comment (agramfort) [inline mne/dipole.py]: a min_dist='auto' would be nice indeed. For this PR or a next one.

for now the urgent need is a test

**Comment by agramfort** [comment]
@jmontoyam see #2171 for how to add tests. Ping we when you added tests. You can just push to this branch to update the PR.

**Comment by agramfort** [comment]
@jmontoyam don't forget me :)

**Comment by agramfort** [comment]
@jmontoyam last one here :)

**Comment by jmontoyam** [comment]
Hi @agramfort, @Eric89GXL,

I have added the test, but something weird is happening, the test fails, here you can see the reason:

---- Fitted :     0.0 ms, distance to inner skull : 4.9270 mm
1 dipoles fitted

The distance of the estimated dipole to the inner skull is  4.9270 mm, but the minimum distance
specify in fit_dipole is 5mm, therefore it does not fulfil the restriction!. 
I think the main reason of this weird behaviour is some numerical instability, but I cannot find the line of code which is causing this numerical instability. Could you please give me some advice?

Thank you very much in advance! :)

File: mne/dipole.py (hunk 1 above)
Comment (agramfort) [inline mne/dipole.py]: @jmontoyam why did you remove this?

File: mne/dipole.py (hunk 2 above)
Comment (agramfort) [inline mne/dipole.py]: still todo

File: mne/dipole.py (hunk 1 above)
Comment (agramfort) [inline mne/dipole.py]: ok forget it. It's below. Sorry for the noise.

File: mne/dipole.py (hunk 4 above)
Comment (agramfort) [inline mne/dipole.py]: todo

File: mne/dipole.py (hunk 6 above)
Comment (agramfort) [inline mne/dipole.py]: +1

File: mne/dipole.py [hunk 7]
```diff
@@ -553,9 +581,11 @@ def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
...
-    out = _fit_dipoles(data, times, src['rr'], guess_fwd_svd, fwd_data,
+    out = _fit_dipoles(min_dist_to_inner_skull, data, times, src['rr'],
+                       guess_fwd_svd, fwd_data,
                        whitener, proj_op, n_jobs)
     dipoles = Dipole(times, out[0], out[1], out[2], out[3], comment)
     residual = out[4]
+    dist_to_inner_skull = out[5]
     logger.info('%d dipoles fitted' % len(dipoles.times))
-    return dipoles, residual
+    return dipoles, residual, dist_to_inner_skull
```
Comment (agramfort) [inline mne/dipole.py]: you're changing the API of a public function. You cannot do this. If it's just for the test do differently.

File: mne/tests/test_dipole.py [hunk 8]
```diff
@@ -151,3 +154,57 @@ def test_len_index_dipoles():
...
+    simulated_scalp_map = simulated_scalp_map[:, None]
+
+    min_distance = 5  # distance in mm
+
+    evoked = EvokedArray(simulated_scalp_map, info, tmin=0)
+
+    info['projs'] = []  # get rid of SSP projections
+    if 'eeg' in evoked:
+        evoked = evoked.copy()
+        avg_proj = make_eeg_average_ref_proj(evoked.info)
```
Comment (agramfort) [inline mne/tests/test_dipole.py]: the addition of the average reference should be done automatically in the fit_dipole like we do for the MNE code.

**Comment by agramfort** [comment]
@Eric89GXL if you want to take over go for it. @jmontoyam did most of the job but he needs more experience to finish quickly on this. Otherwise it can wait a bit...

**Comment by jmontoyam** [comment]
@agramfort, @Eric89GXL,
I have modified the code and the test taking into account @agramfort comments.
Please excuse me my beginner mistakes! :)

**Comment by Eric89GXL** [comment]
I'll have to look into how to get it to strictly obey the limit... I seem to remember there is some way to do it.

File: mne/dipole.py [hunk 9]
```diff
@@ -445,11 +470,22 @@ def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
     """
     # This could eventually be adapted to work with other inputs, these
     # are what is needed:
+
+    evoked.info['projs'] = []  # get rid of SSP projections
+    if 'eeg' in evoked:
+        evoked = evoked.copy()
+        avg_proj = make_eeg_average_ref_proj(evoked.info)
+        evoked.add_proj([avg_proj])
+        evoked.apply_proj()
```
Comment (Eric89GXL) [inline mne/dipole.py]: why get rid of existing projs?

I think we should replicate the proj check that is used in minimum norm inverses, which includes a warning about average EEG ref (not forcing it to be there) IIRC

File: mne/tests/test_dipole.py [hunk 10]
```diff
@@ -151,3 +157,80 @@ def test_len_index_dipoles():
...
+    points, faces = read_surface(fname)
+    points = apply_trans(trans['trans'], points * 1e-3)
+
+    pos = dip.pos
+    ori = dip.ori
+
+    from sklearn.neighbors import NearestNeighbors
+    nn = NearestNeighbors()
+    nn.fit(points)
+    depth, idx = nn.kneighbors(pos, 1, return_distance=True)
```
Comment (Eric89GXL) [inline mne/tests/test_dipole.py]: I thought we had a function that did this sort of thing without strictly requiring `sklearn`, no?

**Comment by Eric89GXL** [comment]
I might have time next week to take this over if that would be useful

File: mne/dipole.py [hunk 11]
```diff
@@ -342,13 +342,15 @@ def _fit_Q(fwd_data, whitener, proj_op, B, B2, B_orig, rd):
     return Q, gof, B_residual
 
 
-def _fit_dipoles(data, times, rrs, guess_fwd_svd, fwd_data, whitener,
+def _fit_dipoles(min_dist_to_inner_skull, data, times, rrs, guess_fwd_svd,
+                 fwd_data, whitener,
                  proj_op, n_jobs):
```
Comment (agramfort) [inline mne/dipole.py]: nitpick but arguments would fit on 2 lines only

File: mne/dipole.py [hunk 12]
```diff
@@ -342,13 +342,15 @@ def _fit_Q(fwd_data, whitener, proj_op, B, B2, B_orig, rd):
...
     """Fit a single dipole to the given whitened, projected data"""
     from scipy.optimize import fmin_cobyla
     parallel, p_fun, _ = parallel_func(_fit_dipole, n_jobs)
     # parallel over time points
-    res = parallel(p_fun(B, t, rrs, guess_fwd_svd, fwd_data, whitener, proj_op,
+    res = parallel(p_fun(min_dist_to_inner_skull, B, t, rrs, guess_fwd_svd,
+                         fwd_data, whitener, proj_op,
                          fmin_cobyla)
```
Comment (agramfort) [inline mne/dipole.py]: same here

File: mne/dipole.py [hunk 13]
```diff
@@ -375,7 +378,19 @@ def constraint(rd):
                                         return_dists=True)[1][0]
                 return -dist
             else:
-                return 1.
+
+                # Once we know the dipole is below the inner skull,
+                # let's check if its distance to the inner skull is at least
+                # min_dist_to_inner_skull. If it is not, it is penalized
```
Comment (agramfort) [inline mne/dipole.py]: it is not penalized. It is not a regularization. It's a constraint.

File: mne/dipole.py [hunk 14]
```diff
@@ -407,11 +422,19 @@ def constraint(rd):
     amp = np.sqrt(np.sum(Q * Q))
     norm = 1 if amp == 0 else amp
     ori = Q / norm
+
+    dist_to_inner_skull = _compute_nearest(surf['rr'],
+                                           rd_final[np.newaxis, :],
+                                           return_dists=True)[1][0]
+    logger.info('---- Fitted : %7.1f ms, distance to inner skull : %2.4f mm' %
+                (1000 * t, dist_to_inner_skull * 1000))
```
Comment (agramfort) [inline mne/dipole.py]: 1000 -> 1000.

get used to write floats with .

File: mne/dipole.py [hunk 15]
```diff
@@ -445,11 +470,22 @@ def fit_dipole(evoked, cov, bem, trans=None, n_jobs=1, verbose=None):
     """
     # This could eventually be adapted to work with other inputs, these
     # are what is needed:
+
+    evoked.info['projs'] = []  # get rid of SSP projections
```
Comment (agramfort) [inline mne/dipole.py]: indeed don't get rid of all projections. Just add the EEG avg ref if not present.

@Eric89GXL to me we should add it automatically. Warning is not enough to make sure people are doing it right.

File: mne/tests/test_dipole.py [hunk 16]
```diff
@@ -151,3 +157,80 @@ def test_len_index_dipoles():
...
+    dip, residual = fit_dipole(evoked, cov, fname_bem, fname_trans,
+                               min_dist=min_distance)
+
+    dist, radiality = _compute_depth_and_radiality(dip, fname_trans,
+                                                   subject, subjects_dir)
+
+    assert (dist * 1000) > min_distance
+
+
+def _compute_depth_and_radiality(dip, fname_trans, subject, subjects_dir):
```
Comment (agramfort) [inline mne/tests/test_dipole.py]: we don't need radiality for the test. Get rid of it please.

File: mne/tests/test_dipole.py (hunk 10 above)
Comment (agramfort) [inline mne/tests/test_dipole.py]: maybe but we sklearn for the fit so the test can depend on it.

my 2c

File: mne/tests/test_dipole.py [hunk 17]
```diff
@@ -151,3 +157,80 @@ def test_len_index_dipoles():
...
+
+    # Simulated scal map
+    simulated_scalp_map = np.zeros(picks.shape[0])
+    simulated_scalp_map[27] = 1
+    simulated_scalp_map[28] = 1
+    simulated_scalp_map[29] = 1
+    simulated_scalp_map[30] = 1
+    simulated_scalp_map[31] = 1
+    simulated_scalp_map[32] = 1
+    simulated_scalp_map[33] = 1
```
Comment (agramfort) [inline mne/tests/test_dipole.py]: simulated_scalp_map[[27, 28, ...]] = 1

would be more compact

File: mne/dipole.py (hunk 15 above)
Comment (Eric89GXL) [inline mne/dipole.py]: we should check to see what we do in the inverse code, it seems like we should do the same thing in both places. so if we don't do it there, maybe we should start...?

File: mne/tests/test_dipole.py (hunk 17 above)
Comment (Eric89GXL) [inline mne/tests/test_dipole.py]: and `simulated_scalp_map[27:34]` would be still more compact :)

File: mne/tests/test_dipole.py (hunk 10 above)
Comment (Eric89GXL) [inline mne/tests/test_dipole.py]: For the dipole fit, you mean? that only uses `scipy`. I thought we had a function that looked for nearest neighbors already using either `scipy` or `sklearn` depending on user capability, but maybe we don't. It seems like it could just be a `np.min` of a `cdist`-like operation (which we definitely have somewhere) even if we don't have a nearest-neighbor function, though.

**Comment by jmontoyam** [comment]
@ agramfort , @Eric89GXL,

at the beginning of the fit_dipole function I add the EEG avg ref if not present, but if it is present but not active, I need to apply only this projector. Currently it is only possible to apply all projectors to the data, but we need to apply only one. To achieve this I did a small hack, what do you think about it?...do you know another clean way to achieve this?

**Comment by jmontoyam** [comment]
@agramfort, 
I misspell your id :) in the last comment, in the case you did not receive it, I hereby send you a ping :), please read the previous comment.

**Comment by agramfort** [comment]
closing in favor of #2196
